nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/QueryDatabaseTable.java

```diff
@@ -0,0 +1,607 @@
+             final Statement st = con.createStatement()) {
+
+            // Try a query that returns no rows, for the purposes of getting metadata about the columns. It is possible
+            // to use DatabaseMetaData.getColumns(), but not all drivers support this, notably the schema-on-read
+            // approach as in Apache Drill
+            String query = getSelectFromClause(tableName, maxValueColumnNames).append(" WHERE 1 = 0").toString();
+            ResultSet resultSet = st.executeQuery(query);
+            ResultSetMetaData resultSetMetaData = resultSet.getMetaData();
+            int numCols = resultSetMetaData.getColumnCount();
+            if (numCols > 0) {
+                columnTypeMap.clear();
+                for (int i = 1; i <= numCols; i++) {
+                    String colName = resultSetMetaData.getColumnName(i).toLowerCase();
+                    int colType = resultSetMetaData.getColumnType(i);
+                    columnTypeMap.put(colName, colType);
+                }
+
+            } else {
+                throw new ProcessException("No columns found in table from those specified: " + maxValueColumnNames);
+            }
+
+        } catch (SQLException e) {
+            throw new ProcessException(e);
+        }
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile fileToProcess = null;
+        if (context.hasIncomingConnection()) {
+            fileToProcess = session.get();
+
+            // If we have no FlowFile, and all incoming connections are self-loops then we can continue on.
+            // However, if we have no FlowFile and we have connections coming from other Processors, then
+            // we know that we should run only if we have a FlowFile.
+            if (fileToProcess == null && context.hasNonLoopConnection()) {
+                return;
+            }
+        }
+
+        final ProcessorLog logger = getLogger();
+
+        final DBCPService dbcpService = context.getProperty(DBCP_SERVICE).asControllerService(DBCPService.class);
+        final String tableName = context.getProperty(TABLE_NAME).getValue();
+        final String columnNames = context.getProperty(COLUMN_NAMES).getValue();
+        final String maxValueColumnNames = context.getProperty(MAX_VALUE_COLUMN_NAMES).getValue();
+        final String preProcessStrategy = context.getProperty(SQL_PREPROCESS_STRATEGY).getValue();
+        final StateManager stateManager = context.getStateManager();
+        final StateMap stateMap;
+        final MaxValueResultSetRowCollector maxValCollector = new MaxValueResultSetRowCollector(stateManager, preProcessStrategy);
+
+        try {
+            stateMap = stateManager.getState(Scope.CLUSTER);
+        } catch (final IOException ioe) {
+            getLogger().error("Failed to retrieve observed maximum values from the State Manager. Will not perform "
```

Comment:
This is done (but on the next line so this note still appears in the updated diff :)